### ISSUE TITLE
Class Labelling Issues

### ISSUE THREAD

**Issue by justinbt1** [wellcometrust/grant_hrcs_tagger#10]
## Labelling Issues
I think I've found multiple issues that cause the strange labelling behaviour we see in the metrics! 

### Overview
The `split_data_frame` function in the `preprocess.py` module uses the Sklearn MultiLabelBinarizer to convert the feature labels into a one-hot-encoded format.

The code in this pull request:
```python
mlb = MultiLabelBinarizer()
y = mlb.fit_transform(df[category])

...

train = pd.DataFrame(y_train, columns=mlb.classes_)
test = pd.DataFrame(y_test, columns=mlb.classes_)
```
The MultiLabelBinarizer orders the one-hot-encoded binary columns and returns the class labels in lexicographical order. See this issue https://github.com/scikit-learn/scikit-learn/issues/4450

The `preprocess.py` module saves the class labels and associated occurrence counts for each class to a JSON file called `value_counts.json`. The dictionary is generated from a Pandas `.value_counts()` method, which orders the classes in order of occurrence.

The value_counts JSON class labels and occurrence counts are then used in the `plot_metrics` function of the `train.py` module to log the class labels to Weights and Biases, along with their scores and counts. This is problematic as the labels and counts in the `value_counts` dict are ordered by occurrence count while the metrics are ordered lexicographically from the MultiLabelBinarizer in `preprocess.py`, meaning the labels are not paired with the correct metric scores.

An attempt to sort the metrics into the same order as the labels and counts is made in `plot_metrics`:
```python
# order metrics to align with value_counts index
value_count_keys = list(value_counts.keys())
metric_idx = [value_count_keys.index(label) for label in value_count_keys]
f1 = [f1[idx] for idx in metric_idx]
precision = [precision[idx] for idx in metric_idx]
recall = [recall[idx] for idx in metric_idx]
```
However all this does is iterate over each list, keeping all of them in the original order.

This is due to the below code, which simply plots the position of each key in value_counts in order, essentially from $0$ to $n$, without changing the order:
```python
value_count_keys = list(value_counts.keys())
metric_idx = [value_count_keys.index(label) for label in value_count_keys]
```
Effectively the equivalent of:
```python
metric_idx = [*range(0, len(value_counts))]
```
Which means for example the below simply moves the F1 in position 0 to 0, 1 to 1, and so on...
```python
f1 = [f1[idx] for idx in metric_idx]
```
Even if this re-ordering did work, it would still be incorrect as the correct order is not provided.  
This incorrectly paired data is then logged to Weights and Biases.

**Notes**:
- I don't think this affects the performance of the model, though it does limit it's evaluation.
- These issues have existed since the first pull request and should be taken into account for model evaluations that post-date those changes.

### Proposed Fix
- Remove value counts JSON and associated code, instead use dataframe column names.
- Calculate counts in `train.py` using the Pandas group_by method.

**Comment by ArneRobben** [comment]
Thanks very good points. 

The original idea for integrating the value_counts was to be able to compare F1 scores per class to the number of examples we have of that class, to get a sense if certain classes were underperforming simply because of underrepresentation. This is a pretty useful for evaluating. Given that the binarizer is lexicographical, would this simple fix work?

```
metric_idx = [sorted(value_count_keys).index(label) for label in value_count_keys]
```
